src/stateless_validation.cpp

```diff
@@ -1,48 +1,240 @@
+#include <algorithm>
+#include <string>
+#include <vector>
+
 #include "veriblock/arith_uint256.hpp"
 #include "veriblock/consts.hpp"
 #include "veriblock/stateless_validation.hpp"
+#include "veriblock/strutil.hpp"
 
 namespace VeriBlock {
 
 static const ArithUint256 MAXIMUM_DIFFICULTY(
     "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF");
 
-bool checkProofOfWork(const BtcBlock& block) {
+bool checkBitcoinTransactionForPoPData(const VbkPopTx& tx,
+                                       ValidationState& state) {
+  WriteStream stream;
+  tx.publishedBlock.toRaw(stream);
+  tx.address.getPopBytes(stream);
+
+  // finding that stream data contains in the tx.bitcoinTransaction
+  for (size_t i = 0, j = 0;
+       i < tx.bitcoinTransaction.tx.size() - stream.data().size() + 1;
+       ++i) {
+    bool found = true;
+    for (; j < stream.data().size(); ++j) {
+      if (tx.bitcoinTransaction.tx[i + j] != stream.data()[j]) {
+        found = false;
+        break;
+      }
+    }
+    if (found) {
+      return true;
+    }
+  }
+
+  return state.Invalid(
+      "checkBitcoinTransactionForPoPData()"
+      "Invalid Vbk Pop transaction",
+      "Bitcoin transaction does not contain PoP publication data");
+}
+
+bool checkBitcoinBlocks(const std::vector<BtcBlock>& btcBlock,
+                        ValidationState& state) {
+  uint256 lashHash;
+  for (const auto& block : btcBlock) {
+    if (!checkBtcBlock(block, state)) {
+      return state.addStackFunction("checkBitcoinBlocks()");
+    }
+
+    if (lashHash != uint256()) {
```

Comment:
I think it is better to initialize lastHash with the first block. Make sure to check for empty blocks list before taking first block.

Reply:
fixed